EIP5007: TimeNFT, ERC-721 Time Extension (#5007)
ethereum/EIPs: EIPS/eip-5007.md

---
eip: 5007
title: TimeNFT, ERC-721 Time Extension
description: Add start time and end time to ERC-721 tokens.
author: Anders (@0xanders), Lance (@LanceSnow), Shrug <shrug@emojidao.org>
discussions-to: https://ethereum-magicians.org/t/eip5007-erc721-time-extension/8924
status: Draft
type: Standards Track
category: ERC
created: 2022-04-13
requires: 165, 721
---

## Abstract

> This standard is an extension of [ERC-721](./eip-721.md). It proposes some additional properties( `startTime`, `endTime`,`originalTokenId`) to help with on-chain time management.

Reviewer: ```suggestion
This standard is an extension of [ERC-721](./eip-721.md). It proposes some additional properties (`startTime`, `endTime`,`originalTokenId`) to help with on-chain time management.
```

## Motivation

Some NFTs have a defined usage period and cannot be used outside of that period. With traditional NFTs that do not include time information, if you want to mark a token as invalid or enable it at a specific time, you need to actively submit a transaction—a process both cumbersome and expensive.

Some existing NFTs contain time functions, but their interfaces are not consistent, so it is difficult to develop third-party platforms for them.

> By introducing (`startTime`, `endTime`) and unifying the naming, it is possible to enable and disable NFT automatically on chain.

Reviewer: ```suggestion
By introducing these properties (`startTime`, `endTime`) and unifying the naming, it is possible to enable and disable NFT automatically on chain.
```

---
eip: 5007
title: TimeNFT, ERC-721 Time Extension
description: Add start time and end time to ERC-721 tokens.
author: Anders (@0xanders), Lance (@LanceSnow), Shrug <shrug@emojidao.org>
discussions-to: https://ethereum-magicians.org/t/eip5007-erc721-time-extension/8924
status: Draft
type: Standards Track
category: ERC
created: 2022-04-13
requires: 165, 721
---

## Abstract

> This standard is an extension of [ERC-721](./eip-721.md). It proposes some additional property( `startTime`, `endTime`,`originalTokenId`) to help with the on-chain time management.

Reviewer: ```suggestion
This standard is an extension of [ERC-721](./eip-721.md). It proposes some additional properties (`startTime`, `endTime`,`originalTokenId`) to help with on-chain time management.
```

## Motivation

> Some NFTs have a defined usage period and cannot be used when they are not at a specific time. If you want to make NFT invalid when it is not in use period, or make NFT enabled at a specific time, while the NFT does not contain time information, you often need to actively submit the chain transaction, this process is both cumbersome and a waste of gas.

Reviewer: Could you include an example of what kind of NFTs have a time component?

Reviewer 2: > Could you include an example of what kind of NFTs have a time component?

1) [ENS](https://etherscan.io/address/0x57f1887a8bf19b14fc0df6fd9b2acc9af147ea85#readContract) :
nameExpires(id) of ENS   -->  endTime(id) of  EIP-5007 ;  
register time of ENS -->  startTime(id) of EIP-5007 ；

2) [doNFT](https://etherscan.io/address/0x8B81e64b6d4ebE2EC0f155df28c54aE2F20189DD#readContract) of Double Protocol: 
getDuration(id)[0] , start time of the lease   --> startTime(id) of EIP-5007 
getDuration(id)[1] , end time of the lease--> endTime(id) of EIP-5007

Reviewer: Sorry, I meant in the EIP itself :rofl:

> Some NFTs have a defined usage period and cannot be used when they are not at a specific time. If you want to make NFT invalid when it is not in use period, or make NFT enabled at a specific time, while the NFT does not contain time information, you often need to actively submit the chain transaction, this process is both cumbersome and a waste of gas.

Reviewer: ```suggestion
Some NFTs have a defined usage period and cannot be used outside of that period. With traditional NFTs that do not include time information, if you want to mark a token as invalid or enable it at a specific time, you need to actively submit a transaction—a process both cumbersome and expensive.
```

> There are also some NFTs contain time functions, but the naming is different, third-party platforms are difficult to develop based on it.

Reviewer: ```suggestion
Some existing NFTs contain time functions, but their interfaces are not consistent, so it is difficult to develop third-party platforms for them.
```

By introducing (`startTime`, `endTime`) and unifying the naming, it is possible to enable and disable NFT automatically on chain.

## Specification

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

```solidity
interface IERC5007  {

    /// @notice Emitted when the `startTime` or `endTime` of a NFT is changed
    /// @param tokenId  The tokenId of the NFT
    /// @param startTime  The new start time of the NFT
    /// @param endTime  The new end time of the NFT

>     event TimeUpdate(uint256 tokenId, uint64 startTime, uint64 endTime);

Reviewer: Is this emitted when tokens are minted/burned? Is the behaviour in constructors different?

Reviewer 2: > Is this emitted when tokens are minted/burned? Is the behaviour in constructors different?

`TimeUpdate` will not be emitted when tokens are minted/burned. I will change this event to `MetadataUpdate` of [EIP4906](https://github.com/ethereum/EIPs/pull/4906)

    /// @notice Get the start time of the NFT
    /// @dev Throws if `tokenId` is not valid NFT
    /// @param tokenId  The tokenId of the NFT
    /// @return The start time of the NFT

>     function startTime(uint256 tokenId) external view returns (uint64);

Reviewer: If `block.timestamp == IERC5007.startTime(0x...)`, is the token valid or invalid?

Reviewer 2: > If `block.timestamp == IERC5007.startTime(0x...)`, is the token valid or invalid?

valid.

    
    /// @notice Get the end time of the NFT
    /// @dev Throws if `tokenId` is not valid NFT
    /// @param tokenId  The tokenId of the NFT
    /// @return The end time of the NFT

>     function endTime(uint256 tokenId) external view returns (uint64);

Reviewer: If `block.timestamp == IERC5007.endTime(0x...)`, is the token valid or invalid?

Reviewer: Also note this too. I know it's in the comment for `isValidNow`, but it could be more visible.

    /// @notice Get the token id which this NFT mint from
    /// @dev Throws if `tokenId` is not valid NFT
    /// @param tokenId  The tokenId of the NFT
    /// @return The token id which this NFT mint from

>     function originalTokenId(uint256 tokenId) external view returns (uint256);

Reviewer: I don't understand what this does from this comment alone.

I'd guess that if you had a token `0x01`, and you split it into `0x01` and `0x02`, then `originalTokenId(0x02) == 0x01`?

What happens if you then merged `0x02` with `0x03`, which happened to be adjacent?

Reviewer 2: > I'd guess that if you had a token `0x01`, and you split it into `0x01` and `0x02`, then `originalTokenId(0x02) == 0x01`?

Yes,  `originalTokenId(0x02) == 0x01`

Reviewer: > What happens if you then merged `0x02` with `0x03`, which happened to be adjacent?

I'm still curious about this case.

    /// @param tokenId  The tokenId of the NFT
    /// @return The the NFT is valid now
    /// if(startTime <= now <= endTime) {return true;} else {return false;}
    function isValidNow(uint256 tokenId) external view returns (bool);

    /// @notice Mint a new token from an old token
    /// @dev Throws if `tokenId` is not valid token
    /// @param originalTokenId_  The token id which the new token mint from
    /// @param newTokenStartTime  The start time of the new token
    /// @param newTokenOwner  The owner of the new token
    /// @return newTokenId The the token id of the new token
    function split(uint256 originalTokenId_, uint64 newTokenStartTime, address newTokenOwner) external returns(uint256);

    /// @notice Merge two time NFTs into one time NFT
    /// @dev Throws if `firstTokenId` or `secondTokenId` is not valid token
    /// @param firstTokenId   The id of the first token
    /// @param secondTokenId  The id of the second token
    /// @param newTokenOwner  The owner of the new token
    /// @return newTokenId The id of the new token
    function merge(uint256 firstTokenId, uint256 secondTokenId, address newTokenOwner) external returns(uint256);
}
```

> 

Reviewer: You'll want to include the EIP-165 interface id, and require that implementers implement `supportsInterface`

## Rationale

### Time Data Type

> The max value of `uint64` is 18446744073709551615,  timestamp 18446744073709551615 is about year 584942419325. `uint64` is sufficient and saves gas compared to `uint256`.

Reviewer: Does this actually save gas? I thought all integer types encoded to 32 bytes: https://docs.soliditylang.org/en/develop/abi-spec.html#examples

Reviewer 2: > Does this actually save gas? I thought all integer types encoded to 32 bytes: 

``` struct TimeNftInfo {  uint256 originalTokenId;  uint64 startTime;  uint64 endTime; }```  encoded to 64 bytes.

``` struct TimeNftInfo {  uint256 originalTokenId;  uint256 startTime;  uint256 endTime; }```  encoded to 96 bytes.

Reviewer 2: Another reason，`uint256` is too big for some software  such as MySQL、Elastic Search.  `uint64` is natively supported on many software and programming languages.

Reviewer 3: > Another reason，uint256 is too big for some software such as MySQL, Elastic Search. uint64 is natively supported on many software and programming languages.

That's still completely arbitrary.

I just ran some tests on a project that already makes heavy use of time-based NFTs to confirm whether it was even worth it to switch to uint64 for storage. Our gas costs actually ended going up...

I highly recommend switching to uint256 completely.

Reviewer 2: > I just ran some tests on a project that already makes heavy use of time-based NFTs to confirm whether it was even worth it to switch to uint64 for storage. Our gas costs actually ended going up...

Could you please send the code link of this  time-based NFTs?

Reviewer 3: It's private at the moment, but the reasoning for can be elaborated on. There is still some minimal amount of arithmetic and comparisons against other uint256's, and the uint64 => uint256 transitions adds gas everytime. For example, a common use-case is to add or compare against *now* (`block.timestamp`) which is natively `uint256`.

## Backwards Compatibility

> As mentioned in the specifications section, this standard can be fully ERC721 compatible by adding an extension function set.

Reviewer: ```suggestion
As mentioned in the specifications section, this standard can be fully ERC-721 compatible by adding an extension function set.
```

    }

    /// @notice mint a new original time NFT
    /// @param to_  The owner of the new token
    /// @param startTime_  The start time of the new token
    /// @param endTime_  The end time of the new token
    /// @return newTokenId The id of the new token
    function _mintOriginalToken(address to_, uint64 startTime_, uint64 endTime_) internal virtual returns(uint256 newTokenId) {
        newTokenId = _nextTokenId;
        _nextTokenId++;
    
        TimeNftInfo storage info = _timeNftMapping[newTokenId];
        info.originalTokenId = newTokenId;
        info.startTime = startTime_;
        info.endTime = endTime_;

        _mint(to_, newTokenId );
    }

    /// @notice burn a time NFT
    /// @param tokenId  The id of the token
    function _burn(uint256 tokenId) internal  virtual override{
        super._burn(tokenId);
        delete _timeNftMapping[tokenId];
    }

    /// @dev See {IERC165-supportsInterface}.
    function supportsInterface(bytes4 interfaceId) public view virtual override returns (bool) {
        return interfaceId == type(IERC5007).interfaceId || super.supportsInterface(interfaceId);
    }
}

```

## Security Considerations

> Implementors of the `ERC5007` standard must consider the condition of `merge` function.  The `originalTokenId` of two NFTs should be same and the time of two NFTs should be adjacent.

Reviewer: I don't understand what this means.

Reviewer 2: > 

I will add more test code later.

Reviewer: I meant more the English description of the security consideration. I don't understand what you're trying to convey. I think it just needs to be reworded a bit to be more understandable.

    function startTime(uint256 tokenId) external view returns (uint64);
    
    /// @notice Get the end time of the NFT
    /// @dev Throws if `tokenId` is not valid NFT
    /// @param tokenId  The tokenId of the NFT
    /// @return The end time of the NFT
    function endTime(uint256 tokenId) external view returns (uint64);

}
```

The `supportsInterface` method MUST return `true` when called with `0x7a0cdf92`.

## Rationale

### Time Data Type

The max value of `uint64` is 18446744073709551615,  timestamp 18446744073709551615 is about year 584942419325. `uint256` is too big for some software such as MySQL, Elastic Search, and `uint64` is natively supported on mainstream programming languages.

## Backwards Compatibility

As mentioned in the specifications section, this standard can be fully ERC-721 compatible by adding an extension function set.

## Test Cases

Test cases are included in [test.js](../assets/eip-5007/test/test.js). 

See [README.md](../assets/eip-5007/README.md) for how to run the test cases.
 
## Reference Implementation
See [ERC5007.sol](../assets/eip-5007/contracts/ERC5007.sol).

## Security Considerations

No security issues found.

## Copyright

> Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

Reviewer 4: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE).
```
